File: problems-typescript/1261.ts


class TreeNode {
    val: number
    left: TreeNode | null
    right: TreeNode | null
    constructor(val?: number, left?: TreeNode | null, right?: TreeNode | null) {
        this.val = (val===undefined ? 0 : val)
        this.left = (left===undefined ? null : left)
        this.right = (right===undefined ? null : right)
    }
}

class FindElements {
    public tgtMap: Set<number>
    constructor(root: TreeNode | null) {
        this.tgtMap = new Set<number>();
        if (root != null) {
            root.val = 0;
            this.tgtMap.add(root.val);
            this.build(root);
        }
    }
    build(root: TreeNode | null) {
        if (root.left != null) {
            root.left.val = 2 * root.val + 1;
            this.tgtMap.add(2 * root.val + 1);
            this.build(root.left);
        }
        if (root.right != null) {
            root.right.val = 2 * root.val + 2;
            this.tgtMap.add(2 * root.val + 2);
            this.build(root.right);
        }
    }
    find(target: number): boolean {
        return this.tgtMap.has(target);
    }
}

File: problems-cs/1261.cs


using System.Collections.Generic;

public class TreeNode
{
    public int val;
    public TreeNode? left;
    public TreeNode? right;
    public TreeNode(int val = 0, TreeNode? left = null, TreeNode? right = null)
    {
        this.val = val;
        this.left = left;
        this.right = right;
    }
}

public class FindElements
{
    public HashSet<int> hset = new HashSet<int>();

    public FindElements(TreeNode root)
    {
        root.val = 0;
        hset.Add(root.val);
        Build(root);
    }

    public void Build(TreeNode root)
    {
        if (root.left != null)
        {
            root.left.val = root.val * 2 + 1;
            hset.Add(root.left.val);
            Build(root.left);
        }
        if (root.right != null)
        {
            root.right.val = root.val * 2 + 2;
            hset.Add(root.right.val);
            Build(root.right);
        }
    }

    public bool Find(int target)
    {
        return this.hset.Contains(target);
    }
}